### PR TITLE
[collectionsjs] Added type for 

### PR DESCRIPTION
**Pull request by jaymeh**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/collectionsjs/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// Type definitions for collectionsjs 0.3
+// Project: https://github.com/logaretm/collectionsjs#readme
+// Definitions by: Jaymeh <https://github.com/jaymeh>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.1
+
+export default class Collection {
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: This seems like a prime candidate for a generic eg. `export default class Collection<T>` and replace all(where applicable) usage of `any` with T.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/collectionsjs/index.d.ts (hunk 1 above)
Comment (jaymeh) [reply types/collectionsjs/index.d.ts]: Thanks for pointing this out, I'm rather new to typescript so this concept is a little confusing but I've now revised my request, hopefully correctly. 👍

File: types/collectionsjs/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    all(): Collection<T>;
+    average(property?: string | ((property?: number) => number)): number;
+    chunk(size: number): Collection<T>;
+    collect(collectable: T[]): Collection<T>;
+    concat(collection: T[]|Collection<T>): Collection<T>;
+    contains(closure: ((item: T) => boolean)): boolean;
+    count(): number;
+    each(callback: (item: T) => void): Collection<T>;
+    filter(callback: (item: T) => boolean): Collection<T>;
+    find(item: any): number;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: `item: T`

File: types/collectionsjs/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    first(callback?: ((item: T) => boolean)|null): T;
+    flatten(deep?: boolean): Collection<T>;
+    get(index: number): T;
+    has(item: T): boolean;
+    join(separator?: string): string;
+    keys(): Collection<T>;
+    last(callback?: ((item: T) => boolean)|null): T;
+    map(callback: (item: T) => any): Collection<T>;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: `map<R>(callback (item: T) => R): Collection<R>`

File: types/collectionsjs/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    pluck(property: string): Collection<T>;
+    push(item: T): Collection<T>;
+    reduce(callback: (previous: T, current: T) => any, initial: any): any;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: reduce<R>(reduce(callback: (previous: R, current: T) => R, initial: R): R;`

File: types/collectionsjs/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    reject(callback: (item: T) => boolean): Collection<T>;
+    remove(item: any): boolean;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: `item: T`

File: types/collectionsjs/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    reverse(): Collection<T>;
+    skip(count: number): Collection<T>;
+    slice(start: number, end?: number): Collection<T>;
+    sort(compare?: () => boolean): Collection<T>;
+    sortBy(property: string, order?: string): Collection<T>;
+    stringify(): string;
+    sum(property?: string|null): any;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: Probably:
```
sum(property: T extends object ? keyof T : never): number
```
or something, not sure.

File: types/collectionsjs/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    take(count: number): Collection<T>;
+    macro(name: string, callback: (...args: any) => any): any;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]: `static macro(name: string, callback: (coll: Collection<unknown>, ...args: unknown[]) => unknown): void`

File: types/collectionsjs/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    unique(callback?: string|null|((item: T) => any)): Collection<T>;
+    values(): Collection<T>;
+    where(callback: ((item: T) => boolean)|string, value?: any): Collection<T>;
```
Comment (SimonSchick) [inline types/collectionsjs/index.d.ts]:
```
where<K extends keyof T>(key: K, value: T[K]): Collection<T>
where(callback: (item: T) => boolean): Collection<T>
```

**Review by SimonSchick — CHANGES_REQUESTED** [review]
Some more notes, TS is a beast :)

**Comment by jaymeh** [comment]
Thanks for the above comments on this. I have implemented what was mentioned above. Is there any way you could offer a little bit of input specifically surrounding the "macro" functionality and how to implement it in my definition? Essentially it is used to add extra functions to the class by amending the prototype allowing someone to add their own functions outside of our definition.

I asked on [stack overflow](https://stackoverflow.com/questions/55004479/documenting-and-allowing-for-macro-function-on-javascript-class?noredirect=1#comment96768183_55004479) earlier today for some advice on it but haven't got any useful responses.

**Comment by SimonSchick** [comment]
@jaymeh TS doesn't really have any way to declare runtime defined method as it only knows what going on compile/transpile time.

The only thing you can do is do a per-`macro` definitions augmentation.

**Comment by jaymeh** [comment]
@SimonSchick Ok, fair point. Is that something I need to do specifically as part of my PR or is it that new definitions of macros will be typed by the creator?

**Comment by SimonSchick** [comment]
You can't really do this as a part of this PR.

**Review by SimonSchick — APPROVED** [review]

**Pull request merged**